[docs] Add bugfix release criteria (#5707)
open-telemetry/opentelemetry-collector: docs/release.md

1. `unknown revision internal/coreinternal/v0.55.0` -- This is typically an indication that there's a dependency on a new module. You can fix it by adding a new `replaces` entry to the `go.mod` for the affected module. 

## Bugfix releases

### Bugfix release criteria

Both `opentelemetry-collector` and `opentelemetry-collector-contrib` have very short 2 week release cycles. Because of this, we put a high bar when considering making a patch release, to avoid wasting engineering time unnecessarily.

When considering making a bugfix release on the `v0.N.x` release cycle, the bug in question needs to fulfill the following criteria:

> 1. The bug was introduced on the `v0.N-1.x` release cycle.

Reviewer: Does this mean in order to release a `0.55.1` patch, the bug had to appear in `0.55-1.0`, which is `0.54.0`?  Am I interpreting this correctly?

Author: Yes, that's it. Do you think I should phrase it differently?

Reviewer: I think I was expecting a patch release to deal with issues that appeared in the release itself.  Like if 0.55.0 was released in a state such that the collector wouldn't even start, I was expecting a 0.55.1 release immediately, not waiting for 0.56.0 and then 0.56.1.

Combined with criteria #2, it sounds like in order for an issue from 0.55.0 to get patched, it has to be found on Wednesday/Thursday, Friday, or Saturday after the 0.55.0 release, then wait 11 more days for a 0.56.0 release, and then wait for a 0.56.1 release.

Should the first criteria be "The bug was introduced in version `v0.N-1.x` or higher, up to and including the current release"?

Author: Right, sorry, I meant `v0.N.0`, I don't know why I wrote that :smile:

> I was expecting a patch release to deal with issues that appeared in the release itself.

This is what I wanted to write, let me rephrase that

Author: Replaced `N-1`by `N`, hopefully that's clear enough

2. The bug has been reported within the first 3 working days after the official binaries were released.
3. The bug has no workaround or the workaround is significantly harder to put in place than updating the version. Examples of simple workarounds are:
    - Reverting a feature gate.
    - Changing the configuration to an easy to find value.
4. The bug happens in common setups. To gauge this, maintainers can consider the following:

>     - The bug is not specific to a platform that is not widely used

Reviewer: ```suggestion
    - The bug is not specific to an uncommon platform
```
The double "not"s meesed me up lol

    - The bug happens with the default configuration or with a commonly used one (e.g. has been reported by multiple people)
5. The bug is sufficiently severe. For example (non-exhaustive list):
    - The bug makes the Collector crash reliably
    - The bug makes the Collector fails to start under an accepted configuration
    - The bug produces significant data loss
    - The bug makes the Collector negatively affect its environment (e.g. significantly affects its host machine)

The OpenTelemetry Collector maintainers will ultimately have the responsibility to assess if a given bug fulfills all the necessary criteria.

> Exceptions may be granted in a case-by-case basis if there is widespread consensus that a bugfix release should be done despite the above rules.

Author: Should this line be here?

Reviewer: I like the wiggle room, but I think ultimately the "case-by-case" basis is based on the maintainer's judgement.  Could modify the line above to be "The OpenTelemetry Collector maintainers will ultimately have the responsibility to assess if a given bug fulfills all the necessary criteria, or should be patched anyways".

Author: Reworded
